Fsharp Stroustrup style guide (#33171)
dotnet/docs: docs/fsharp/style-guide/formatting.md

Indentation of conditionals depends on the size and complexity of the expressions that make them up.
Write them on one line when:

> * `cond`, `e1`, and `e2` are short

Reviewer: ```suggestion
* `cond`, `e1`, and `e2` are short.
```

Reviewer: Line 649 is a sentence fragment, although I realize you didn't change it here.

        [| 1; 8; 28; 56; 70; 56; 28; 8; 1 |] 
    |]
```

If a list or array expression is the right-hand side of a binding, you may prefer to use "Stroustrup" style:

```fsharp
// ✔️ OK
let pascalsTriangle = [| 
   [| 1 |]
   [| 1; 1 |]
   [| 1; 2; 1 |]
   [| 1; 3; 3; 1 |]
   [| 1; 4; 6; 4; 1 |]
   [| 1; 5; 10; 10; 5; 1 |]
   [| 1; 6; 15; 20; 15; 6; 1 |]
   [| 1; 7; 21; 35; 35; 21; 7; 1 |]
   [| 1; 8; 28; 56; 70; 56; 28; 8; 1 |] 
|]
```

However, when a list or array expression is _not_ the right-hand side of a binding, such as when it's inside of another list or array, if that inner expression needs to span multiple lines, the brackets should go on their own lines:

```fsharp
// ✔️ OK - The outer list follows "Stroustrup" style, while the inner lists place their brackets on separate lines
let fn a b = [ 
    [
        someReallyLongValueThatWouldForceThisListToSpanMultipleLines
        a
    ]
    [ 
        b
        someReallyLongValueThatWouldForceThisListToSpanMultipleLines 
    ]
]

// ❌ Not okay
let fn a b = [ [
    someReallyLongValueThatWouldForceThisListToSpanMultipleLines
    a
]; [
    b
    someReallyLongValueThatWouldForceThisListToSpanMultipleLines
] ]

> ```

Reviewer 2: Semi-related is elmish.  How should we go about `function list1 list2`? I've seen it typically like:

```fsharp
button [
    ClassName ("btn " + if model.NewBook.Validate() then "btn-primary" else "btn-disabled")
    OnClick (fun _ -> dispatch ValidateBook) ] [
    i [ClassName "glyphicon glyphicon-plus"; Style [PaddingRight 5]] []
    str "Add"
]
```

where the `] [` is on the same line.

Reviewer 3: I think in function applications you can justify the `] [` placement as it does not depend on the `;` being present.

Author: I'm not super familiar with the intricacies of elmish code, having never used it. I'd welcome suggestions on how to address that style here

Reviewer 3: We could address this in a future PR, though related, it is a bit of a subject of its own.
I think the rules which Fantomas applies today to format this are a good start.

      Lackeys = ["Zippy"; "George"; "Bungle"] }
```

#### Multiline bracket formatting styles

For records that span multiple lines, there are three commonly used formatting styles: `Cramped`, `Aligned`, and `Stroustrup`. The `Cramped` style has been the default style for F# code, as it tends to favor styles that allow the compiler to easily parse code. Both `Aligned` and `Stroustrup` styles allow for easier reordering of members, leading to code that may be easier to refactor, with the drawback that certain situations may require slightly more verbose code.

* `Cramped` Record style:
The historical standard, and default F# record format. Opening brackets go on the same line as the first member, closing bracket on the same line as the last member.

    ```fsharp

>     let rainbow = 

Reviewer 3: I don't think these code blocks need to be indented.

Author: Even though they're part of the list items?

Reviewer 3: Oh, that's why. Well maybe, I don't know, it will be up to the people reviewing this.
It could be either way in this case.

let point2 = { point with X = 1; Y = 2 }
```

Longer expressions should use new lines, and format based on one of the above-named conventions:

```fsharp
// ✔️ OK - Cramped
let newState =
    { state with
        Foo =
            Some
                { F1 = 0
                  F2 = "" } }
        
// ✔️ OK - Aligned
let newState = 
    {
        state with
            Foo =
                Some
                    {
                        F1 = 0
                        F2 = ""
                    }
    }

// ✔️ OK - Stroustrup
let newState = { 
    state with
        Foo =
            Some { 
                F1 = 0
                F2 = ""
            }
}
```

> **Note**: If using `Stroustrup` style for copy-and-update expressions, members *must* be indented further than the copied record name

Reviewer: ```suggestion
**Note**: If using `Stroustrup` style for copy-and-update expressions, you *must* indent members further than the copied record name:
```

        Lackeys = [ "Zippy"; "George"; "Bungle" ] }
```

You may want to dedicate separate lines for the braces and indent one scope to the right with the expression:

```fsharp
// ✔️ OK
let newState =
    {
        state with
            Foo =
                Some 
                    {
                        F1 = 0
                        F2 = ""
                    }
    }
```

You may also prefer the "Stroustrup" style of keeping the opening brace on the same line as the binding: 

```fsharp
let newState =
    { state with
        Foo =

>             Some { // TODO: Fantomas doesn't respect this yet

Reviewer 3: The examples shouldn't mention what Fantomas does not respect (yet).
Once this is merged, we can follow up on Fantomas' side.

Author: Haha yeah anything marked `TODO` was mainly a note for myself haha, never planned to include those in the final draft. In this case I meant that as a note to myself to open an issue in fantomas if there isn't one already.

              reversed.CompareTo (rev s2) }
```

You may also prefer to use "Stroustrup" style:

```fsharp
let comparer =

>     { new IComparer<string> with

Reviewer 3: Would this not be
```fsharp
let comparer = { 
    new IComparer<string> with
        member x.Compare(s1, s2) =
            let rev (s: String) = new String(Array.rev (s.ToCharArray()))
            let reversed = rev s1
            reversed.CompareTo(rev s2)
}
```
?

We don't seem to support this in Fantomas today.
If there is no reason to believe this will ever lead to invalid code, I guess we can have this as well.

Author: Yeah for some of these suggestions I did limit what I put in to what fantomas supports currently, but I agree this is the ideal here and I'll update it accordingly.

### Formatting let and member declarations

When formatting `let` and `member` declarations, typically the right-hand side of a binding either goes on one line, or (if it's too long) goes on a new line indented one level.

> For example, the following are compliant:

Reviewer: ```suggestion
For example, the following examples are compliant:
```

        printfn "%A" x
```

These are non-compliant: 

```fsharp
// ❌ Not OK, code formatters will reformat to the above by default
let a = """
foobar, long string
"""

let d = while f do
    printfn "%A" x
```

Record type instantiations may also place the brackets on their own lines:

```fsharp
// ✔️ OK
let c =
    { 
        Name = "Bilbo"
        Age = 111
        Region = "The Shire" 
    }
```

You may also prefer to use "Stroustrup" style, with the opening `{` on the same line as the binding name:

```fsharp

> // ✔️ OK

Reviewer 3: Maybe a good section to point out that `Stroustrup` has certain limitations.
For example:
```fsharp
let c = {
    bilbo with 
    Name = "Bilbo"
    Age = 111
    Region = "The Shire" 
}
```
This leads to `Possible incorrect indentation: this token is offside of context started at position`.
However, in that example, if we further indent the update record fields:
```fsharp
let c = {
    bilbo with 
        Name = "Bilbo"
        Age = 111
        Region = "The Shire" 
}
```
it does become valid again.

// ❌ Not OK, code formatters will reformat to the above by default
type PostalAddress =
    { Address: string
      City: string
      Zip: string }
  with
    member x.ZipAndCity = $"{x.Zip} {x.City}"
  end
  
// ✔️ OK
type PostalAddress =
    { 
        Address: string
        City: string
        Zip: string 
    }
    member x.ZipAndCity = $"{x.Zip} {x.City}"
    
// ❌ Not OK, code formatters will reformat to the above by default
type PostalAddress =
    { 
        Address: string
        City: string
        Zip: string 
    }
    with
        member x.ZipAndCity = $"{x.Zip} {x.City}"
    end
```

The exception to this style rule is if you format records according to the "Stroustrup" style. In this situation, due to compiler rules, the `with` keyword is required if you want to implement an interface or add additional members:

```fsharp
// ✔️ OK
type PostalAddress = {
    Address: string
    City: string
    Zip: string
} with
    member x.ZipAndCity = $"{x.Zip} {x.City}"
   
// ❌ Not OK, this is currently invalid F# code
type PostalAddress = {
    Address: string
    City: string
    Zip: string
} 
member x.ZipAndCity = $"{x.Zip} {x.City}"
```

> When XML documentation is added for record fields, it becomes normal to indent and add additional whitespace between members:

Reviewer 3: @josh-degraw what should happen when you have a record type with XML doc and Stroustrup?
I feel like that should also go to `Aligned` of one of the fields has XML.
Thoughts?

Author: Personally, from a user perspective I definitely wouldn't expect it to need to change to aligned. It makes sense to not allow Cramped style there, but I think it should be seen as fine to have xml comments on the type members.

Reviewer 3: Cramped indeed has that problem where it would be on the same line as the XML doc comment.
Stroustrup doesn't have this issue, so might be reasonable to not change it.

### Formatting record declarations

> For record declarations, by default you should indent `{` in type definition by four spaces, start the field list on the same line and align any members with the `{` token:

Reviewer: ```suggestion
For record declarations, by default you should indent `{` in type definitions by four spaces, start the field list on the same line, and align any members with the `{` token:
```

```fsharp
// ✔️ OK
type PostalAddress =
    { Address: string
      City: string
      Zip: string }
```

It is also common to prefer putting brackets on their own line, with members indented by an additional four spaces:

```fsharp
// ✔️ OK
type PostalAddress =
    { 
        Address: string
        City: string
        Zip: string 
    }
```

> You may also put the `{` at the end of the type declaration line like ("stroustrup" style):

Reviewer: ```suggestion
You may also put the `{` at the end of the type declaration line ("Stroustrup" style):
```

```fsharp
// ✔️ OK
type PostalAddress = {
    Address: string
    City: string
    Zip: string
}
```

> If additional members are needed, whenever possible, don't use `with`/`end`:.

Reviewer: ```suggestion
If additional members are needed, don't use `with`/`end` whenever possible:
```

    interface IMyInterface
```

> These same rules apply for anonymous record type aliases:

Reviewer: ```suggestion
These same rules apply for anonymous record type aliases.
```

The domain that's being modeled should ultimately drive the naming convention.
If it's idiomatic to use a different convention, that convention should be used instead.

If the return value of an expression is a computation expression, prefer putting the computation expression keyword name on its own line

```fsharp
// ✔️ OK
let foo () = 
    async {
        let! value = getValue()
        do! somethingElse()
        return! anotherOperation value 
    }
    

> // ❌ Not OK, but formatters may allow this if preferred // TODO Should these just both be acceptable styles?

Reviewer 3: Yeah, I guess both styles are ok. I'm not sure how to phrase this. You want to add something that people shouldn't mix the two styles in teams.

    end
```

> // TODO: Should Stroustroup be allowed for inline anonymous record types here, and if so should there be any restrictions/exceptions?

Reviewer 3: I'm not aware of any limitations here, I think this should be fine to add.
